Intelligently convert C/C++ comments to Rust

dylanmckay opened the pull request:
With this change, we can correctly parse C++ block comments.

```cpp
/**
 * Does a thing
 *
 * More documentation. This test does something
 * useful.
 */
```

into

```rust
/// Does a thing
///
/// More documentation. This test does something
/// useful.
```

Fixes servo/rust-bindgen#426.

dylanmckay commented:
Note that this does not implement the `///<` doxygen syntax (which causes the doc comment to apply to the previous item).

As far as I can tell, this is currently already broken and is causing us to emit the wrong doc comments off by one on struct fields/enum variants which use the syntax.

dylanmckay commented:
I need to modify this so that it also treats regular comments (`//`) as doc comments, will do so later today.

emilio commented:
This is lovely, thank you!

Could we also add the test-case from #426?

dylanmckay commented:
I can do that.

Also, I'm going to refactor this so that rather than doing the C++ -> Rust comment conversion at the codegen level, we do it when building the AST. As far as I understand it, that would be a more appropriate place and (depending on how we build the AST - I haven't looked at it yet), we may be able to have a single call which converts comments rather than having multiple over the place.

emilio commented:
Yeah, I think that's fine, though if it gets nasty we could also move the `preprocess` call to `attributes::doc`, and use it where it's not currently used. That'd remove the duplication too.

dylanmckay commented:
I've added the test case from #426 and also moved the preprocessing to the IR stage.

Let me know if you'd like me to change anything in order to merge :)

emilio commented:
Thanks again for doing this btw! :)

dylanmckay commented:
> 

> Thanks again for doing this btw! :)

No problem, thanks for working on the project :)


I've fixed the comment issue but I can't get the `convert-cpp-comments-to-rust` test I added to match the expected header file.

Here is a part of the diff from the [travis run](https://travis-ci.org/servo/rust-bindgen/jobs/250748065)

It seems as if it's related to environment differences between my computer and CI, but I can't really replicate the exact same build setup as them.

```
pub const _STDINT_H: ::std::os::raw::c_uint = 1;
 pub const _FEATURES_H: ::std::os::raw::c_uint = 1;
+pub const __USE_ANSI: ::std::os::raw::c_uint = 1;
 pub const _ISOC95_SOURCE: ::std::os::raw::c_uint = 1;
 pub const _ISOC99_SOURCE: ::std::os::raw::c_uint = 1;
-pub const _ISOC11_SOURCE: ::std::os::raw::c_uint = 1;
 pub const _POSIX_SOURCE: ::std::os::raw::c_uint = 1;
 pub const _POSIX_C_SOURCE: ::std::os::raw::c_uint = 200809;
 pub const _XOPEN_SOURCE: ::std::os::raw::c_uint = 700;
 pub const _XOPEN_SOURCE_EXTENDED: ::std::os::raw::c_uint = 1;
 pub const _LARGEFILE64_SOURCE: ::std::os::raw::c_uint = 1;
-pub const _DEFAULT_SOURCE: ::std::os::raw::c_uint = 1;
+pub const _BSD_SOURCE: ::std::os::raw::c_uint = 1;
+pub const _SVID_SOURCE: ::std::os::raw::c_uint = 1;
 pub const _ATFILE_SOURCE: ::std::os::raw::c_uint = 1;
-pub const __USE_ISOC11: ::std::os::raw::c_uint = 1;
 pub const __USE_ISOC99: ::std::os::raw::c_uint = 1;
 pub const __USE_ISOC95: ::std::os::raw::c_uint = 1;
 pub const __USE_POSIX: ::std::os::raw::c_uint = 1;
 pub const __USE_POSIX2: ::std::os::raw::c_uint = 1;
 pub const __USE_POSIX199309: ::std::os::raw::c_uint = 1;
 pub const __USE_POSIX199506: ::std::os::raw::c_uint = 1;
 pub const __USE_XOPEN2K: ::std::os::raw::c_uint = 1;
 pub const __USE_XOPEN2K8: ::std::os::raw::c_uint = 1;
 pub const __USE_XOPEN: ::std::os::raw::c_uint = 1;
 pub const __USE_XOPEN_EXTENDED: ::std::os::raw::c_uint = 1;
 pub const __USE_UNIX98: ::std::os::raw::c_uint = 1;
 pub const _LARGEFILE_SOURCE: ::std::os::raw::c_uint = 1;
 pub const __USE_XOPEN2K8XSI: ::std::os::raw::c_uint = 1;
 pub const __USE_XOPEN2KXSI: ::std::os::raw::c_uint = 1;
 pub const __USE_LARGEFILE: ::std::os::raw::c_uint = 1;
 pub const __USE_LARGEFILE64: ::std::os::raw::c_uint = 1;
 pub const __USE_MISC: ::std::os::raw::c_uint = 1;
+pub const __USE_BSD: ::std::os::raw::c_uint = 1;
+pub const __USE_SVID: ::std::os::raw::c_uint = 1;
```

Does this ring any bells @emilio?

emilio commented on tests/headers/convert-cpp-comment-to-rust.hpp:
```diff
@@ -0,0 +1,16 @@
+#include <stddef.h>
+#include <stdint.h>
```
Yeah, tests can't have system includes to pass, since these are machine-dependent.

emilio commented on tests/headers/convert-cpp-comment-to-rust.hpp:
```diff
@@ -0,0 +1,16 @@
+#include <stddef.h>
+#include <stdint.h>
+
+typedef uint32_t mbedtls_mpi_uint;
```
instead you can remove them and, let's say, replace `uint32_t` for `unsigned`, and `size_t` for `unsigned long`. they're irrelevant to the test-case anyway, so it seems sensible.

dylanmckay replied:
Very helpful, thanks!

dylanmckay commented:
Tests are now green @emilio, thanks for the help!

emilio commented:
Looks great, thanks again for this!

@bors-servo r+

bors-servo commented:
:pushpin: Commit 239a015 has been approved by `emilio`

<!-- @bors-servo r=emilio 239a0154cdab5fa0052f2ce98129e2c169dc1cc4 -->

bors-servo commented:
:hourglass: Testing commit 239a0154cdab5fa0052f2ce98129e2c169dc1cc4 with merge f19e576961d2e3832760ca22a8e62f8ab4f7f078...

bors-servo commented:
:sunny: Test successful - [status-travis](https://travis-ci.org/servo/rust-bindgen/builds/251516886?utm_source=github_status&utm_medium=notification)
Approved by: emilio
Pushing f19e576961d2e3832760ca22a8e62f8ab4f7f078 to master...

jethrogb commented:
Author of #426 here. I didn't get a notification when this PR was opened.

Why are you converting into Rust doc comments? It seems highly unlikely that `rustdoc` or `rustdoc --test` will be able to deal with random C++ comments correctly.

emilio commented:
> Why are you converting into Rust doc comments? It seems highly unlikely that rustdoc or rustdoc --test will be able to deal with random C++ comments correctly.

Well, sure, but that's followup work anyway. I think it's nice to have the comment in a processable format, which is what this PR enables. This PR also fixes the most common cases where people doesn't use markdown for their C/C++ docs.

If we want to support fully rustdoc, etc, there may be more followup work needed. An easy way to generate comments and make rustdoc not complain is making comments normal `//` comments.

Anyway, I've followed up in #799 with fixes and a few features that this PR enabled. I think Markdown/Doxygen support can be discussed in a new issue. Will file one to that effect.

emilio commented:
I opened #800 for that.

dylanmckay commented:
@emilio would it be possible for a new revision to be published so I can use this in one of my own crates?

emilio commented:
@dylanmckay sure, though I'd wait for #799 to merge to fix the panic when a class is documented with an empty multiline comment (`/**/`), I don't want to release known panics. I'll push a version bump to that PR :)

dylanmckay commented:
Sounds good to me, thanks @emilio :)
